File: app/src/main/java/net/aquadx/aquacard/data/ScoreFormat.kt
package net.aquadx.aquacard.data

import java.util.Locale

/**
 * Чистые null-safe форматтеры кодировок maimai (docs/aquadx-api-spec.md §5).
 * Доступ к позиционным кортежам best* через getOrNull — устойчивость к дрейфу длины.
 * Числовые форматы — Locale.US, чтобы разделитель был точкой независимо от локали.
 */
object ScoreFormat {

    private val MAI_LEVELS = listOf("BASIC", "ADVANCED", "EXPERT", "MASTER", "Re:MASTER")

    /** achievement = % × 10000. 1008790 -> "100.8790%". */
    fun achievementPercent(achievement: Int?): String {
        if (achievement == null) return "—"
        val whole = achievement / 10000
        val frac = achievement % 10000
        return String.format(Locale.US, "%d.%04d%%", whole, frac)
    }

    fun levelName(level: Int?): String {
        if (level == null) return "?"
        return MAI_LEVELS.getOrNull(level) ?: "LV$level"
    }

    /** Буква-ранг maimai по achievement (порог 100.5% = 1005000 и ниже). */
    fun maiRank(achievement: Int?): String = when {
        achievement == null -> "—"
        achievement >= 1005000 -> "SSS+"
        achievement >= 1000000 -> "SSS"
        achievement >= 995000 -> "SS+"
        achievement >= 990000 -> "SS"
        achievement >= 980000 -> "S+"
        achievement >= 970000 -> "S"
        achievement >= 940000 -> "AAA"
        achievement >= 900000 -> "AA"
        achievement >= 800000 -> "A"
        else -> "—"
    }

    fun comboLabel(comboStatus: Int?): String? = when (comboStatus) {
        1 -> "FC"; 2 -> "FC+"; 3 -> "AP"; 4 -> "AP+"; else -> null
    }

    fun syncLabel(syncStatus: Int?): String? = when (syncStatus) {
        1 -> "FS"; 2 -> "FS+"; 3 -> "FDX"; 4 -> "FDX+"; else -> null
    }

    /** maimai rating — целое как есть (16666). */
    fun formatRating(rating: Int?): String = rating?.toString() ?: "—"

    /** Название песни по musicId с фолбэком на сам id, если метаданных нет. */
    fun songName(meta: Map<Int, MusicMeta>, musicId: Int): String =
        meta[musicId]?.name?.takeIf { it.isNotBlank() } ?: musicId.toString()

    /**
     * Парс позиционного кортежа Best maimai: [musicId, levelIndex, v3, achievement] (4 элемента).
     * Любая нехватка элементов -> null, без исключений (getOrNull).
     */
    fun parseBestTuple(tuple: List<String>): BestEntry? {
        val musicId = tuple.getOrNull(0)?.toIntOrNull() ?: return null
        val level = tuple.getOrNull(1)?.toIntOrNull() ?: 0
        val value = tuple.getOrNull(3)?.toIntOrNull() ?: return null
        return BestEntry(musicId, level, value)
    }

    /** Текст значения Best: achievement как процент. */
    fun bestValueLabel(value: Int): String = achievementPercent(value)
}

File: iosApp/AquaCard/ScoreFormat.swift
import Foundation

enum ScoreFormat {
    private static let maiLevels = ["BASIC", "ADVANCED", "EXPERT", "MASTER", "Re:MASTER"]

    static func achievementPercent(_ achievement: Int?) -> String {
        guard let achievement else { return "-" }
        let whole = achievement / 10_000
        let fraction = achievement % 10_000
        return String(format: "%d.%04d%%", locale: Locale(identifier: "en_US_POSIX"), whole, fraction)
    }

    static func levelName(_ level: Int?) -> String {
        guard let level else { return "?" }
        guard maiLevels.indices.contains(level) else { return "LV\(level)" }
        return maiLevels[level]
    }

    static func maiRank(_ achievement: Int?) -> String {
        guard let achievement else { return "-" }
        switch achievement {
        case 1_005_000...: return "SSS+"
        case 1_000_000...: return "SSS"
        case 995_000...: return "SS+"
        case 990_000...: return "SS"
        case 980_000...: return "S+"
        case 970_000...: return "S"
        case 940_000...: return "AAA"
        case 900_000...: return "AA"
        case 800_000...: return "A"
        default: return "-"
        }
    }

    static func comboLabel(_ status: Int?) -> String? {
        switch status {
        case 1: return "FC"
        case 2: return "FC+"
        case 3: return "AP"
        case 4: return "AP+"
        default: return nil
        }
    }

    static func syncLabel(_ status: Int?) -> String? {
        switch status {
        case 1: return "FS"
        case 2: return "FS+"
        case 3: return "FDX"
        case 4: return "FDX+"
        default: return nil
        }
    }

    static func formatRating(_ rating: Int?) -> String {
        rating.map(String.init) ?? "-"
    }

    static func songName(_ meta: [Int: MusicMeta], _ musicId: Int) -> String {
        let name = meta[musicId]?.name?.trimmingCharacters(in: .whitespacesAndNewlines)
        return name?.isEmpty == false ? name! : "\(musicId)"
    }

    static func parseBestTuple(_ tuple: [String]) -> BestEntry? {
        guard tuple.count > 3,
              let musicId = Int(tuple[0]),
              let value = Int(tuple[3]) else {
            return nil
        }
        return BestEntry(musicId: musicId, level: Int(tuple[1]) ?? 0, value: value)
    }

    static func bestValueLabel(_ value: Int) -> String {
        achievementPercent(value)
    }
}

enum AquaAssets {
    static func staticBase(_ baseURL: String) -> String {
        let trimmed = baseURL.trimmingCharacters(in: .whitespacesAndNewlines)
            .removingTrailingSlash()
        if trimmed.hasSuffix("/aqua") {
            return String(trimmed.dropLast(5))
        }
        return trimmed
    }

    static func allMusicURL(_ baseURL: String) -> String {
        "\(staticBase(baseURL))/d/mai2/00/all-music.json"
    }

    static func jacketURL(baseURL: String, musicId: Int) -> URL? {
        let file = String(format: "%06d", musicId % 10_000)
        return URL(string: "\(staticBase(baseURL))/d/mai2/music/\(file).png")
    }

    static func avatarURL(baseURL: String, profilePicture: String?) -> URL? {
        guard let picture = profilePicture?.trimmingCharacters(in: .whitespacesAndNewlines),
              !picture.isEmpty else {
            return nil
        }
        return URL(string: "\(staticBase(baseURL))/uploads/net/portrait/\(picture)")
    }
}

private extension String {
    func removingTrailingSlash() -> String {
        var value = self
        while value.hasSuffix("/") {
            value.removeLast()
        }
        return value
    }
}
